Update Sign In flow for devices with multiple facilities

nucleogenesis opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

### Summary
<!--
 * description of the change
 * manual verification steps performed
 * screenshots if the PR affects the UI
-->

Revision of login page and flow for better multiple facility handling.

Fixes #5875 

### Reviewer guidance
<!--
 * how can a reviewer test these changes?
 * are there any risky areas that deserve extra testing
-->

Design Notes:

- "Go back" will have arrow when KDS is merged in
- Button styles will also need fiddling but not much when KDS is merged in

You'll likely need to test this by creating one or a few new `KOLIBRI_HOME` folders. Before running the server, set `KOLIBRI_HOME=/nonexistant/path` and run the server. I should note that I had an issue where the first attempt at starting failed but I think that might be local - will test that separately anyway. If you have that issue, just run the server again and it should be good.

**Single Facility**

When you only have one facility - there is one design change to test.

- During Setup Wizard, select the "Personal" type of installation. 
- On the sign in page - nothing changes.

- On a new KOLIBRI_HOME - do setup wizard and select Formal, give it a name
- On the sign in page, you should see "Sign into FACILITY NAME"

- Do that again for a Non-Formal installation.
- On the sign in page, you should see "Sign into FACILITY NAME"

When you have multiple facilities, the following all should be true:


> **Note!!!**
> Sign in as your superuser and go to http://localhost:8000/api/auth/facility/ to create a new facility


- Follows the following design flow: https://www.figma.com/file/LM0YUkbWAOIerZ3EEg0FN6bS/P2P-Sync?node-id=605%3A1769
- When you are on the page prompting you to create a new account or sign in - you are taken to the "Select a facility" page.
- You can go back from that page to the previous page.

> **Note!!!!!!**
> Once you've selected a facility, you will not be taken back to the sign-in-or-sign-up prompt page with the two buttons. You need to remove the Local Storage value for 'facilityId' to see that page again.


- When you don't have a cookie set - you will always be taken first to the 'sign-in-or-sign-up' prompt page. Test this by trying to go to http://localhost:8000/en/user/#/signin once you've removed your LocalStorage.

- Test creating a user for a specific facility - then make sure you cannot log in as that user on a different facility and _can_ login as that user on the proper facility.

- Test that you can change your facility from the sign in page.


### References
<!--
 * references to related issues and PRs
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
-->

https://www.notion.so/learningequality/Technical-design-90b9dac261ab4e97bd68b0840cd42a25

----

### Contributor Checklist


PR process:

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [x] Contributor has fully tested the PR manually
- [x] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

nucleogenesis committed:
Initial pass of frontend remodeling

nucleogenesis committed:
lintless styling clean up (see notes)

- buttons look funny because I'm accounting for KDS changes
- arrows missing on "Go back" due ot KDS changes

nucleogenesis committed:
Sign up facility id in submission

nucleogenesis committed:
Pass smoke test - add mock for facilityId value needed

nucleogenesis committed:
Button styling - wrong string shown on faciltyselect

jonboiser commented on kolibri/plugins/user/assets/src/views/SignInPage.vue:
```diff
@@ -2,192 +2,100 @@
-              class="guest small-text"
+    <AuthBase>
+      <div
+        v-if="hasMultipleFacilities && selectedFacility.name"
```
This "Sign into 'Facility name'" should be enabled for a single-facility device when the `DeviceSetting.preset` is formal/nonformal. Right? I tried to take a shortcut by modifying the DeviceSetting model directly instead of going through the setup wizard

nucleogenesis replied:
`selectedFacility` will only have a `name` if it's Formal/Nonformal

However using the preset makes sense - will need to see where that lives in the Store - I only looked for it on the Facility dataset properties

nucleogenesis replied:
That `preset` is set for the Facility. I added it to the `FacilitySerializer` and check for that here instead now.

nucleogenesis replied:
Actually - since this is not shown for an Informal facility - wouldn't we want a separate string altogether here in that case?

If all three facilities presets are possible on a single device - then we still want to show the user the ability to change their facility, right?

I'll raise this in the peer-sync Slack as well - just wanted to note it here.

jonboiser commented on kolibri/plugins/user/assets/src/views/AuthBase.vue:
Hm, is this 'guest access' setting on both Device and Facility settings? I think it might be (it's "Allow users to access resouces..." on the Device Settings page). For this page, we only want to show this link if the Device Setting for guest access is on.

I think we might have left the facility field intact by accident. I see now that it's not something that is configurable for Facilities on the UI anymore.

nucleogenesis replied:
Ahh okay good to know. That's something that was there before but good to change it now since I'm working on the code. I'll update it to use the device setting here

nucleogenesis replied:
I have this fixed and ready to commit but the `GET devicesettings` endpoint permissions need to be changed. I've noted this in #kolibri-dev to see if anybody objects to opening the permissions there and with general approval I'll push it.

jonboiser commented on kolibri/plugins/user/assets/src/views/SignUpPage.vue:
```diff
@@ -40,12 +40,12 @@
           />
 
-          <template v-if="currentFacility">
+          <template v-if="selectedFacility.name">
```
Need to add a "Sign in with an existing account" escape link at the bottom of this form. (See last step of Scenario 1 on Figma)

jonboiser commented on kolibri/plugins/user/assets/src/views/UserIndex.vue:
```diff
@@ -37,12 +41,18 @@
       immersiveProperties() {
         if (this.pageName === PageNames.SIGN_UP) {
           if (!this.$route.query.step) {
+            const backRoute =
```
One fix for Navigation: 

If you are this point

<img width="422" alt="Screen Shot 2020-04-09 at 3 19 44 PM" src="https://user-images.githubusercontent.com/10248067/78945602-94892680-7a75-11ea-9793-ac0fd12301f2.png">

And you click "Create Account"

When you click the "X" exit link

Then you return to the SignIn page. (Currently it takes you to the signin/signup page).

jonboiser commented on kolibri/plugins/user/assets/src/views/FacilitySelect.vue:
```diff
@@ -0,0 +1,105 @@
+    },
+    methods: {
+      setFacility(facilityId) {
+        console.log(`setting facId ${facilityId}`);
```
Should remove this console log

nucleogenesis replied:
I wonder why linting didn't get mad about that, thanks!

jonboiser commented on kolibri/plugins/user/assets/src/views/SignInPage.vue:
```diff
@@ -2,192 +2,100 @@
+        </transition>
+        <transition name="textbox">
+          <KTextbox
+            v-if="needPasswordField"
```
The `learner_can_login_with_no_password` doesn't seem to be reflected when I switch between facilities with differing settings. Maybe the facility settings need to be reloaded when you switch?

nucleogenesis replied:
This was a great catch thanks JB!

nucleogenesis committed:
Explore without account link on SignUp

nucleogenesis committed:
Misc fixes per @jonboiser's feedback

- Nav to sign-up goes back to sign-in (not auth select)
- Faciilty config reloaded on facility change
- console.log removed

nucleogenesis committed:
Add `preset` to FacilityDatasetSerializer, use it

- Check against the preset to determine whether we show the
"Sign into NAME" and change facility links

nucleogenesis committed:
All MFD signinpage has "change" facility link

nucleogenesis committed:
Bootstrap guest access setting in Learn & User &
use it

nucleogenesis committed:
backlint

jonboiser committed:
Fix import order

jonboiser commented on kolibri/plugins/user/assets/src/views/FacilitySelect.vue:
```diff
@@ -0,0 +1,106 @@
+        <div
+          v-for="facility in facilityList['disabled']"
+          :key="facility.id"
+          class="disabled facility-name"
```
I don't think this `.disabled` class is needed. It should display as unstyled text even without it.

jonboiser commented on kolibri/plugins/user/assets/src/views/SignInPage.vue:
```diff
@@ -2,192 +2,100 @@
-              class="guest small-text"
+    <AuthBase>
+      <div
+        v-if="hasMultipleFacilities"
```
I think the logic should be 

1. Show the "Sign in..." prompt whenever there are more than 1 facilities _or_ this is a non-personal instance.
2. Show the "change" link only if there are 1+ facilities.

jonboiser committed:
Remove .disabled class

jonboiser committed:
Update logic for showing “Sign in…” prompt and “change” link

jonboiser committed:
Remove FacilityModal.vue

jonboiser reviewed (approved):
Looks good to me!

jonboiser committed:
Fix guard for informal (aka personal) preset

jonboiser committed:
Make getting selectedFacility.dataset safer
